submissions/trainperspective/trainperspective.md

```diff
@@ -0,0 +1,55 @@
+
+{% include youtube.html id="5nRnnfKMZKQ" %}
+
+Those who downloaded the mod linked in the description know that this was an April first joke. For those who didn’t figure it out, sorry.
+A few people have asked me for details on how I made that video, so here they are.
+
+I’ve seen some people speculating that I somehow extracted 3D models from the game.
+That’s not needed, it’s a simple rotation of what’s already shown on screen.
+Well the idea might be simple, the execution is a bit more complicated.
+
+The first problem you come across is that screens aren’t circles.
+When rotating, stuff gets cut off, while other things are missing.
+
+{% include image.html src='rotation.jpg' alt='Map view of rail network, showing only an eighth of the outposts.' caption='Simply rotating causes problems' href='rotation.jpg’' %}
+
+I ended up applying my usual solution to recording problems for factorio: use screenshots instead. Ingame screenshots are not bound by the limitations of normal screen recording. I simply capture a larger area than is normally visible.
+
+The next problem is that the UI also rotates, this is not something we want.
+Screenshots also come to the rescue here. The take_screenshot command has an option called show_gui. So the trick is to take 2 screenshots every tick, one with ui and one without ui. If we take only the parts that differ, then we end up with the ui, which we can superimpose over the screenshot without ui. At least that was the plan. Various video editing problems made this non viable. For example my video editor did not support lossless formats (that I could find, I tried a bunch). Small differences in encoding end up in the ui.
+
+However I found out that in recent factorio versions the ui is no longer fixed to your player position in screenshots. It is always visible no matter what part of the map you take a screenshot of. So I looked for a color far away from other colors that occur in the game. I settled on pink. And I changed the image of some concret to be pure pink. I turned off clouds. And changed up the location of my gui screenshot so only the pink concrete is visible. This way I can get the ui by green screening, or rather pink screening to be more accurate.
+
+![](ui.mp4)
+
+There were a few small problems with that though. Turns out there are transparent parts in the gui: the details panel that shows up when you hover over things with your mouse. It looked purple now. I ended up manually cutting it out of the gui, and nobody seems to have noticed. 
+
+Next I created a small mod that played a sound whenever I started recording screenshots, this way I could sync up the in-game sounds with the screenshots. Similarly to how movie studios use clappers.
+
+The mod also wrote out the value of game.players[1].vehicle.orientation into a text file. I used this to calculate how much rotation is needed, but smoothed out with some splines.
```

Comment:
<img width="738" alt="graph" src="https://user-images.githubusercontent.com/70716783/126550273-a455695d-e20d-4b0d-bc2f-7670a3936a68.png">
Something like this? or do you want something more elaborate? this is a screenshot from inside my video editor.

Reply:
Now that I see it it's not _that_ exciting. If it would be put in I think it would work best if it had clear axis with lables too. But will require you to plot it somewhere else. So it's up to you if you think it's worth the effort.

Reply:
I don't think adding better axis and labels will make it any better, so I'm not going to do it.
The only way I could see it being exciting is if it was interactive and linked with the video playback or something; but that's way too much effort.